legacy/reqcounter/reqcounter.go

```diff
@@ -16,30 +16,83 @@ limitations under the License.
 
 package reqcounter
 
-import "sync"
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
 
 // RequestCounter records the number of HTTP requests to GCR.
 type RequestCounter struct {
-	mutex    sync.Mutex
-	requests uint64
+	mutex    sync.Mutex // Lock to prevent race-conditions with concurrent processes.
+	requests uint64     // Number of HTTP requests since recording started.
+	quota    uint64     // Maximum requests allowed before GCR throttling occurs.
+	since    time.Time  // When the current counter began recording requests.
+}
+
+// log records the number of HTTP requests found and resets the counter.
+func (rc *RequestCounter) log() {
+	// Grab the lock to prevent race-conditions.
+	rc.mutex.Lock()
+	defer rc.mutex.Unlock()
+	// Log the number of requests found.
+	msg := fmt.Sprintf("Since %s there have been %d requests to GCR.", counter.since.Format("01-02-2006 15:04:05"), counter.requests)
+	logrus.Debug(msg)
+	// Warn of GCR scaling when requests exceed quota.
+	rc.warn()
+	// Reset the counter.
+	rc.reset()
+}
+
+// warn only logs a warning if the number of requests has exceeded the quota.
+func (rc *RequestCounter) warn() {
+	if rc.requests > rc.quota {
+		msg := fmt.Sprintf("The GCR quota of \"%d requests per %d min\" has been exceeded.", quotaLimit, measurementWindow/10)
```

Comment:
Create a unique "error-code"-like string to make it easier to search after-the-fact?

QUOTA_EXCEEDED ?